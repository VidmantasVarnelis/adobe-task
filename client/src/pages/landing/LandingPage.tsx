import React, { useEffect } from 'react';
import { Input } from 'components';
import { useLanding } from './hooks/use-landing.hook';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import { PuffLoader } from 'react-spinners';
import {
  Container,
  LogoHeading,
  InnerContainer,
  SpacingContainer,
  Form,
  LoaderContainer,
  RomanNumeralHeading,
  RomanNumeralText,
  CopyBtn,
  RomanNumeralContainer,
} from './landing-page.styles';

const LandingPage: React.FC = () => {
  const {
    onSubmit,
    onChange,
    romanLetters,
    isLoading,
    isError,
    onCopy,
    isFulfilled,
  } = useLanding();

  useEffect(() => {
    if (isError) toast.error('Conversion failed!');
    if (isFulfilled) toast.success('Successfully converted number!');
  }, [isFulfilled, isError]);

  return (
    <Container>
      <ToastContainer />
      <LogoHeading>Roman Numeral Converter</LogoHeading>
      <InnerContainer>
        <SpacingContainer>
          <Form onSubmit={onSubmit} role="form">
            <Input onChange={onChange} placeholder="Type a number" />
          </Form>
          {isLoading ? (
            <LoaderContainer>
              <PuffLoader size={100} />
            </LoaderContainer>
          ) : (
            romanLetters && (
              <RomanNumeralContainer>
                <RomanNumeralHeading>Roman numeral:</RomanNumeralHeading>
                <RomanNumeralText data-testid="heading" onClick={onCopy}>
                  <CopyBtn>Copy</CopyBtn>
                  {romanLetters}
                </RomanNumeralText>
              </RomanNumeralContainer>
            )
          )}
        </SpacingContainer>
      </InnerContainer>
    </Container>
  );
};

export default LandingPage;
