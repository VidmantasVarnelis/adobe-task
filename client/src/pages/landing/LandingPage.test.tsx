import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import LandingPage from './LandingPage';
import { storeWrapper } from 'utils';
import * as hook from './hooks/use-landing.hook';

Object.defineProperty(navigator, 'clipboard', {
  value: {
    writeText: () => jest.fn().mockImplementation(() => Promise.resolve()),
  },
});
jest.mock('store', () => ({
  useAppDispatch: () => jest.fn(),
  useAppSelector: jest.fn().mockImplementation((func) =>
    func({
      landing: {
        state: {
          roman: 'test',
        },
        status: 'rejected',
      },
    }),
  ),
}));

describe('Landing Page Test Suite', () => {
  it('check if page renders correctly', () => {
    const { container } = render(<LandingPage />, { wrapper: storeWrapper });
    expect(container).toMatchSnapshot();
  });
  it('form component test', async () => {
    render(<LandingPage />, { wrapper: storeWrapper });
    const input = screen.getByLabelText('number-input') as HTMLInputElement;
    fireEvent.change(input, {
      target: { value: '23' },
    });
    fireEvent.submit(screen.getByRole('form'));
    expect(input.value).toBe('23');
  });
  it('copy roman numeral event test', () => {
    jest.spyOn(navigator.clipboard, 'writeText');
    render(<LandingPage />, { wrapper: storeWrapper });
    fireEvent.click(screen.getByTestId('heading'));
    expect(navigator.clipboard.writeText).toHaveBeenCalledTimes(1);
    expect(navigator.clipboard.writeText).toHaveBeenCalledWith('test');
  });
  it('test for useEffect - isFulfilled case', () => {
    jest.spyOn(hook, 'useLanding').mockImplementation(() => ({
      romanLetters: '',
      isFulfilled: true,
      isError: false,
      isLoading: false,
      onChange: jest.fn,
      onCopy: jest.fn,
      onClick: jest.fn(),
      onSubmit: jest.fn,
    }));
    render(<LandingPage />, { wrapper: storeWrapper });
  });
  it('test for isLoading case', () => {
    jest.spyOn(hook, 'useLanding').mockImplementation(() => ({
      romanLetters: '',
      isFulfilled: false,
      isError: false,
      isLoading: true,
      onChange: jest.fn,
      onCopy: jest.fn,
      onClick: jest.fn(),
      onSubmit: jest.fn,
    }));
    render(<LandingPage />, { wrapper: storeWrapper });
  });
});
