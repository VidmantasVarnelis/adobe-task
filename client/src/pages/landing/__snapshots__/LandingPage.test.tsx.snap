// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing Page Test Suite check if page renders correctly 1`] = `
<div>
  <div
    class="sc-eCYdqJ khMReP"
  >
    <div
      class="Toastify"
    />
    <h1
      class="sc-papXJ cUmwOC"
    >
      Roman Numeral Converter
    </h1>
    <div
      class="sc-iBkjds gLLwUj"
    >
      <div
        class="sc-jSMfEi UyFka"
      >
        <form
          class="sc-iqcoie bondUQ"
          role="form"
        >
          <div
            class="sc-bczRLJ ehiAyy"
          >
            <button
              class="sc-gsnTZi fskANS"
              role="button"
            >
              <img
                alt="No icon"
                class="sc-dkzDqf dRzneU"
                src=""
              />
            </button>
            <input
              aria-label="number-input"
              class="sc-hKMtZM Fowqy"
              placeholder="Type a number"
            />
          </div>
        </form>
        <div
          class="sc-gKXOVf hybCsZ"
        >
          <h1
            class="sc-jqUVSM eCYDhl"
          >
            Roman numeral:
          </h1>
          <p
            class="sc-kDDrLX kgboSd"
            data-testid="heading"
          >
            <span
              class="sc-crXcEl hpgDSI"
            >
              Copy
            </span>
            test
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
